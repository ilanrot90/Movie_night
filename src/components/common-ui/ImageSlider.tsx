import React, { useCallback, useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';
import { motion } from 'framer-motion';
import usePrevious from 'hooks/usePrevious';

const ImageSlider = styled.div`
	overflow: hidden;
	position: relative;
	display: flex;
	flex-direction: column;
`;

const SliderArrows = styled.div`
	height: 0;
	left: 0;
	position: absolute;
	top: calc(50% - 35px);
	width: 100%;
	z-index: 1;
`;

const SliderIndicator = styled.div`
	position: absolute;
	bottom: 10px;
	left: 50%;
	transform: translateX(-50%);
	display: flex;
	gap: 10px;
`;

const SliderNav = styled.span<{ isActive?: boolean }>`
	width: 10px;
	height: 10px;
	cursor: pointer;
	border-radius: 50%;
	background-color: ${({ theme }) => theme.dark};
	opacity: ${({ isActive }) => (isActive ? 1 : 0.5)};
	transition: opacity ${({ theme }) => theme.utils.quickTransition};
	&:hover {
		opacity: ${({ isActive }) => (isActive ? 1 : 0.75)};
	}
`;

const SliderArrow = styled.span<{ leftArrow?: boolean }>`
	background: none;
	border: none;
	height: 40px;
	opacity: 0.5;
	outline: none;
	position: absolute;
	transition: opacity ${({ theme }) => theme.utils.quickTransition};
	padding: 0;
	width: 21px;
	cursor: pointer;
	&:hover {
		opacity: 1;
	}

	&::after,
	&::before {
		background: ${({ theme }) => theme.dark};
		border-radius: 3px;
		content: '';
		display: block;
		height: 23px;
		position: absolute;
		width: 4px;
		left: 8px;
	}

	&::before {
		top: 0;
	}

	&::after {
		bottom: 0;
	}
	${({ leftArrow }) =>
		leftArrow
			? css`
					top: 0;
					left: 10px;

					&::before {
						transform: rotate(30deg);
					}

					&::after {
						transform: rotate(-30deg);
					}
			  `
			: css`
					top: 0;
					right: 10px;

					&::before {
						transform: rotate(-30deg);
					}

					&::after {
						transform: rotate(30deg);
					}
			  `}
`;

const SliderImage = styled(motion.div).attrs({
	'data-testid': 'image-slider',
	variants: {
		initial: (showNext: boolean) => ({
			x: showNext ? '100%' : '-100%',
		}),
		in: {
			x: 0,
		},
		out: (showNext: boolean) => ({
			x: showNext ? '-100%' : '100%',
		}),
	},
	transition: {
		type: 'spring',
		stiffness: 500,
		damping: 100,
		duration: 0.125,
	},
	initial: 'initial',
	animate: 'in',
	exit: 'out',
})<{ url: string; prevSlide: string; custom: boolean; height: number }>`
	background-image: url(${({ url }) => url});
	background-size: cover;
	background-position: center;
	background-repeat: no-repeat;
	height: ${({ height }) => height}px;
	position: relative;
	&::before {
		background-image: url(${({ prevSlide }) => prevSlide});
		background-size: cover;
		background-position: center;
		background-repeat: no-repeat;
		content: '';
		position: absolute;
		height: 100%;
		width: 100%;
		left: ${({ custom }) => (custom ? -100 : 100)}%;
	}
`;

interface IProps {
	slides: string[];
	height: number;
	autoSlide?: boolean;
	timeFrame?: number;
}

const MAXIMUM_SET_TIMEOUT_VALUE = 2147483647;

const Slider: React.FC<IProps> = ({ slides = [], height, autoSlide = true, timeFrame = 5000 }) => {
	const [[currentSlideIdx, showNext], setCurrentSlide] = useState<[number, boolean]>([0, true]);
	const prevSlideIdx = usePrevious(currentSlideIdx) || 0;
	const timerId = useRef<number>();
	const time = autoSlide ? timeFrame : MAXIMUM_SET_TIMEOUT_VALUE;

	const showNextSlide = useCallback(() => {
		setCurrentSlide(([prevCurrentSlideIdx]) => [(prevCurrentSlideIdx + 1) % slides.length, true]);
	}, [setCurrentSlide, slides.length]);

	const showPrevSlide = useCallback(() => {
		setCurrentSlide(([prevCurrentSlideIdx]) => [prevCurrentSlideIdx === 0 ? slides.length - 1 : prevCurrentSlideIdx - 1, false]);
	}, [setCurrentSlide, slides.length]);

	const showSelectedSlide = useCallback(
		(slideIdx: number) => () => {
			setCurrentSlide(([prevCurrentSlideIdx]) => [slideIdx, prevCurrentSlideIdx < slideIdx]);
		},
		[setCurrentSlide]
	);

	const clearTimer = useCallback(() => {
		clearTimeout(timerId.current);
	}, [timerId]);

	const startTimer = useCallback(() => {
		timerId.current && clearTimer();
		timerId.current = setTimeout(() => {
			showNextSlide();
		}, time);
	}, [showNextSlide, timerId, clearTimer, time]);

	useEffect(() => {
		startTimer();
		return () => {
			clearTimer();
		};
	}, [currentSlideIdx, clearTimer, startTimer]);

	return (
		<ImageSlider onMouseEnter={clearTimer} onMouseLeave={startTimer}>
			<SliderImage
				prevSlide={slides[prevSlideIdx]}
				url={slides[currentSlideIdx]}
				key={currentSlideIdx}
				custom={showNext}
				height={height}
			/>
			<SliderArrows>
				<SliderArrow role="button" onClick={showNextSlide} />
				<SliderArrow role="button" leftArrow onClick={showPrevSlide} />
			</SliderArrows>
			<SliderIndicator>
				{Array.from({ length: slides.length }).map((_, i) => (
					<SliderNav key={i} isActive={i === currentSlideIdx} onClick={showSelectedSlide(i)} />
				))}
			</SliderIndicator>
		</ImageSlider>
	);
};

export default React.memo(Slider);
