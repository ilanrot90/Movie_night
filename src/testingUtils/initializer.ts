import axiosInstance from 'axiosRequest';
import MocksManager from './mocks/axiosMock';
import { setupIntersectionObserverMock, restoreIntersectionObserverMock } from './mocks/IntersectionObserver';

const originalError = console.error;

beforeAll(() => {
	setupIntersectionObserverMock();
	MocksManager.initialize(axiosInstance);
	console.error = (...args: Array<any>) => {
		if (/Warning.*not wrapped in act/.test(args[0])) {
			return;
		}
		originalError.call(console, ...args);
	};
});

beforeEach(() => {
	MocksManager.reset();
});

afterEach(async () => {
	await MocksManager.validate();
});

afterAll(() => {
	console.error = originalError;
	restoreIntersectionObserverMock();
});
