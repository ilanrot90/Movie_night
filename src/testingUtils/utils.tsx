import React, { ComponentType, ReactElement } from 'react';
import userEvent, { TargetElement } from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';
import App from 'App';
import ThemeProvider from 'style/ThemeProvider';
import { RecoilRoot } from 'recoil';
import { MemoryRouter } from 'react-router-dom';
import { fromPairs } from 'utils/lodash.utils';

const GETTERS = ['Text', 'Role', 'TestId', 'LabelText', 'AltText', 'PlaceholderText'] as const;
const gettersMap = {
	Text: 'findByText' as 'findByText',
	Role: 'findByRole' as 'findByRole',
	TestId: 'findByTestId' as 'findByTestId',
	LabelText: 'findByLabelText' as 'findByLabelText',
	AltText: 'findByAltText' as 'findByAltText',
	PlaceholderText: 'findByPlaceholderText' as 'findByPlaceholderText',
};
const gettersKeys = {
	Text: 'clickByText' as 'findByText',
	Role: 'clickByRole' as 'findByRole',
	TestId: 'clickByTestId' as 'findByTestId',
	LabelText: 'clickByLabelText' as 'findByLabelText',
	AltText: 'clickByAltText' as 'findByAltText',
	PlaceholderText: 'clickByPlaceholderText' as 'findByPlaceholderText',
};
const clickByFactory = (fn: Function) => async (...args: Array<any>) => {
	const el: TargetElement = await fn(...args);
	el && userEvent.click(el);
	return el;
};
const changeByFactory = (fn: Function) => async (value: string, ...args: Array<any>) => {
	const el: TargetElement = await fn(...args);
	el && (await userEvent.type(el, value));
	return el;
};

const customQueries = {
	...fromPairs(GETTERS.map(v => [gettersKeys[v], clickByFactory(screen[gettersMap[v]])])),
	...fromPairs(GETTERS.map(v => [gettersKeys[v], changeByFactory(screen[gettersMap[v]])])),
};
// helper to test login screens
export const signOutRender = async () => {
	renderUi({ route: '/app' });
	expect(screen.getByRole('button', { name: /log out/i })).toBeInTheDocument();
	userEvent.click(screen.getByRole('button', { name: /log out/i }));
	return await waitFor(() => expect(screen.getAllByRole('button', { name: /log in/i })).toBeArrayOfSize(4));
};

const renderProviders = (route: string) => ({ children }: { children: ReactElement }): ReactElement => (
	<RecoilRoot>
		<ThemeProvider>
			<MemoryRouter initialEntries={[route]}> {children}</MemoryRouter>
		</ThemeProvider>
	</RecoilRoot>
);
type RenderOptions = {
	route: string;
};
export function renderUi({ route, ...renderOptions }: RenderOptions) {
	return render(<App />, { wrapper: renderProviders(route) as ComponentType, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
export { userEvent, customQueries };
