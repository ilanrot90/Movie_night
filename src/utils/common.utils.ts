export const isProduction = process.env.NODE_ENV === 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const asyncHandler = <T>(promise: Promise<T>) => {
	return promise
		.then(response => ({
			error: undefined,
			response,
		}))
		.catch(error => ({
			response: undefined,
			error,
		}));
};
