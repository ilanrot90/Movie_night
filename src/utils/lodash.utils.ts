import keyBy from 'lodash/keyBy';
import mapValues from 'lodash/mapValues';
import map from 'lodash/map';
import orderBy from 'lodash/orderBy';
import groupBy from 'lodash/groupBy';
import isEmpty from 'lodash/isEmpty';

const chainableFunctions = {
	map,
	orderBy,
	groupBy,
};

const chain = (input: any) => {
	let value = input;
	const wrapper = {
		...mapValues(chainableFunctions, f => (...args: any) => {
			// @ts-ignore
			value = f(...args);
			return wrapper;
		}),
		value: () => value,
	};

	return wrapper;
};

export { chain, map, orderBy, groupBy, keyBy, isEmpty };
