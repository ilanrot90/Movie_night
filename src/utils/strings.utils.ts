export const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
export const nameRegex = /^[a-zA-Z\s]*$/;
export const extendHex = (v: string): string => v.replace(/#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])/, '#$1$1$2$2$3$3');
