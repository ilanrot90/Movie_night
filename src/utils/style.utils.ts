import { extendHex } from './strings.utils';

const isHexColor = (v: string): boolean => /^#([A-Fa-f0-9]{3}){1,2}$/.test(v);

export type HexToRgba = (hex: string, a?: number) => string;
export const hexToRgba: HexToRgba = (hex, a = 1) => {
	if (!isHexColor(hex)) {
		throw new Error('Bad Hex');
	}
	const h = hex.length === 4 ? extendHex(hex) : hex;

	const r = parseInt(h.slice(1, 3), 16);
	const g = parseInt(h.slice(3, 5), 16);
	const b = parseInt(h.slice(5, 7), 16);

	return `rgba(${r}, ${g}, ${b}, ${a})`;
};
