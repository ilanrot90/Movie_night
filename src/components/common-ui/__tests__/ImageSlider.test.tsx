import React from 'react';
import snapshotDiff from 'snapshot-diff';
import { renderUi, waitFor, screen } from 'testingUtils/utils';
import ImageSlider from '../ImageSlider';

const slides = [
	'https://d33wubrfki0l68.cloudfront.net/dd23708ebc4053551bb33e18b7174e73b6e1710b/dea24/static/images/wallpapers/shared-colors@2x.png',
	'https://d33wubrfki0l68.cloudfront.net/49de349d12db851952c5556f3c637ca772745316/cfc56/static/images/wallpapers/bridge-02@2x.png',
	'https://d33wubrfki0l68.cloudfront.net/594de66469079c21fc54c14db0591305a1198dd6/3f4b1/static/images/wallpapers/bridge-01@2x.png',
];

const mockedConsole = jest.spyOn(console, 'error').mockImplementation(() => {});

describe('Render image carousel component', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterAll(() => {
		mockedConsole.mockRestore();
		jest.useRealTimers();
	});

	test('Change image after 3 sec', async () => {
		renderUi({
			app: <ImageSlider slides={slides} height={400} timeFrame={3000} />,
			route: '/',
		});
		const currentSliderImage = screen.getByTestId(/image-slider/i);

		jest.advanceTimersByTime(3001);
		const nextSliderImage = screen.getByTestId(/image-slider/i);
		expect(snapshotDiff(currentSliderImage, nextSliderImage)).toMatchSnapshot();
	});

	test('does not attempt to set state when unmounted (to prevent memory leaks)', () => {
		const { unmount } = renderUi({
			app: <ImageSlider slides={slides} height={400} />,
			route: '/',
		});
		unmount();
		waitFor(() => jest.runOnlyPendingTimers());
		expect(mockedConsole).not.toHaveBeenCalled();
	});
});
