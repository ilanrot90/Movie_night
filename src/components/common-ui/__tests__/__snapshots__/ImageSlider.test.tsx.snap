// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render image carousel component Change image after 3 sec 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <div
    animate=\\"in\\"
-   class=\\"sc-AxgMl dxpndP\\"
+   class=\\"sc-AxgMl lRIwE\\"
    data-testid=\\"image-slider\\"
    exit=\\"out\\"
    height=\\"400\\"
    initial=\\"initial\\"
    prevslide=\\"slide-image-1\\"
    role=\\"image-slider\\"
    transition=\\"[object Object]\\"
-   url=\\"slide-image-1\\"
+   url=\\"slide-image-2\\"
    variants=\\"[object Object]\\"
  />"
`;
